### DISCUSSION TITLE
Inline R code in YAML

### DISCUSSION THREAD

**Discussion by mbojan** [quarto-dev/quarto-cli#606] [Q&A] [answered]
This is related Q to #452 . Is it generally possible to have inline R code in YAML? It worked in RMarkdown. I can't get it to work using either `r 1 + 1` or `!expr 1 + 1`. Am I missing something? Thanks!

**Comment by cderv** [comment]
When using the YAML syntax for chunk option using `#|` inside code blocks, you won't be able to use inline R chunk syntax that **knitr** supports ``` `r 1 + 1` ```.  The special syntex with `!expr` needs to be used ans in the example with `fig-cap` in https://quarto.org/docs/computations/r.html#chunk-options

You'll need last version of Quarto and also it requirees dev version of **knitr** as we have fixed an issue with this recently. 
https://github.com/quarto-dev/quarto-cli/issues/547

**Reply by mbojan** [reply]
Thanks @cderv . I'm interested in evaluating R expression in YAML header, e.g. in title or author, not in chunk options. Is this possible currently? Will this be possible and using which syntax?

**Reply by cderv** [reply]
R expression in YAML header won't work as this is Quarto which is parsing and not R. 

What is your usage exactly regarding programming metadata field ? Do you have an example ?

**Reply by mbojan** [reply]
> R expression in YAML header won't work as this is Quarto which is parsing and not R.

Ah, OK. So that's different from RMarkdown, isn't it.


> What is your usage exactly regarding programming metadata field ? Do you have an example ?

I'm generating HTML reports based on some RDS files and I would like to have titles and some other metadata (e.g. description) containing some info based on the content of these files. I guess can get away with containing these in the first paragraph of the report, but modifying YAML metadata would be much nicer...

**Reply by mbojan** [reply]
I believe [this](https://bookdown.org/yihui/rmarkdown-cookbook/document-metadata.html) mechanism does not work either. Is there anything comparable in Quarto, i.e. are the document metadata accessible computationally?

**Reply by cderv** [reply]
Regarding the access of metadata inside the document, this is possible using a feature called shortcodes 
See https://quarto.org/docs/authoring/shortcodes-and-filters.html#shortcodes for the `meta` shortcode

**Reply by cderv** [reply]
> I'm generating HTML reports based on some RDS files and I would like to have titles and some other metadata (e.g. description) containing some info based on the content of these files.

I think you would need to create a workflow to do that differently than with R Markdown. Using R Markdown, you can do that because R is use as the computation engine with **knitr** but also as the workflow layer than handles the output format, and calls Pandoc. With Quarto this is different: R is only used at the computation engine level. 

However, there is tools to create workflows around rendering documents: Project scripts for example can be useful (https://quarto.org/docs/projects/quarto-projects.html#project-scripts)

Regarding setting title and description without writing the content in the YAML file, for now the solution I see is that you can do that by passing metadata to Pandoc through Quarto I believe. At command line that would look like 
```sh
quarto render test.Rmd -M title:'my new title'
```
Not as ideal though but could be a workaround.

**Reply by cderv** [reply]
Note that you could also use a YAML to set some metadata that you can set in YAML header
````
---
author: "me"
format: html
metadata-files: 
	- include-meta.yml
---

# Hello

World 
````

`include-meta.yml` would be
````yaml
title: My Document
description: An interesting content
````

Then you could write a R script to modify the `include-meta.yml` file content based on your different values, and render the same `.qmd` to get different titles and descriptions.

Hope it helps

**Reply by mbojan** [reply]
> Regarding the access of metadata inside the document, this is possible using a feature called shortcodes See https://quarto.org/docs/authoring/shortcodes-and-filters.html#shortcodes for the `meta` shortcode

@cderv , the shortcodes can only be used to render text and I can't use them in the code chunks, correct?

**Reply by cderv** [reply]
what do you have in mind ? Having examples helps. Shortcodes are handled by Quarto directly, and not by computations engines, but not sure to see what you are trying to achieve. In case you do not know, there is also parameters features in Quarto: https://quarto.org/docs/computations/parameters.html

**Reply by mbojan** [reply]
I'm exploring the features you mentioned to check what I can achieve in terms of 

> > What is your usage exactly regarding programming metadata field ? Do you have an example ?
> 
> I'm generating HTML reports based on some RDS files and I would like to have titles and some other metadata (e.g. description) containing some info based on the content of these files. I guess can get away with containing these in the first paragraph of the report, but modifying YAML metadata would be much nicer...

... without having to resort to external scripts generating files with metadata or executing Quarto CLI with commandline parameters.

**Reply by cderv** [reply]
I don't think you will be able to have the same kind of workflow that with **rmarkdown** where could do some process during knitting that would impact how quarto is working. However, I will be interested to see what workflow you will come up with that would suites you and work with all these. 

As we often remind, there is no reason to switch from R Markdown to Quarto if you are happy with the feature there, and especially if you heavily use R for the workflow. 

> the shortcodes can only be used to render text and I can't use them in the code chunks, correct?

I am still no sure to see clearly how you would like to use shortcodes in code chunk ? Shortcodes should be resolved before computations or after computations in your case ?

**Reply by cderv** [reply]
To add on other feature that works also within Quarto, you can use `var` shortcodes within YAML frontmatter when you are using a Quarto project (https://quarto.org/docs/authoring/shortcodes-and-filters.html#var)
 
This would be 
````
---
title: '{{< var title >}}'
---
````
and then set the value within `_variables.yml` files.

**Comment by cderv** [comment]
@jjallaire with **rmarkdown** it is possible to use **knitr** execution to set some YAML field value programatically for Pandoc to read. Example in https://bookdown.org/yihui/rmarkdown-cookbook/dynamic-yaml.html, either using inline R code 
````
---
title: "An analysis of `r nrow(mtcars)` cars"
---
````

or params

````
---
title: "`r params$doc_title`"
author: "Smart Analyst"
params:
  doc_title: "The Default Title"
---
````

With Quarto we can't do exactly the same as R is only use as computation engine but is there any equivalent mechanism to set title using parameters when generating several reports for example ? 

Only solution I se and mentioned above is setting metadata using `-M` when calling `quarto render`, and can't really be done inside the document.

**Comment by jjallaire** [comment]
You can also pass `--metadata-file` on the command line.

**Reply by cderv** [reply]
Thanks! 
I opened an issue about that as I think it is not working as expected, at least on Windows
https://github.com/quarto-dev/quarto-cli/issues/613

**Reply by mbojan** [reply]
It does not work for me on Linux (WSL) either...

**Comment by mbojan** [comment, marked as answer]
For future generations: from all the discussion above I'm distilling that the metadata in the YAML document header (`title`, `author`, `date`, `description` etc.)...

1. ... can contain the inline R code `` `r 1 + 1` `` but is *ignored*. YAML fields are not passed through **knitr**
2. ... can be accessed in the document "body" via shortcodes to render some text (c.f. [shortcodes](https://quarto.org/docs/authoring/shortcodes-and-filters.html#shortcodes))
3. ... are *not* visible in code chunks in any form (contrary to `metadata` object in **rmarkdown**, c.f. [this](https://bookdown.org/yihui/rmarkdown-cookbook/document-metadata.html))
4. ... can't be modified by the code *within* the `qmd` file
5. ... can be set externally by using `-M`, e.g. `quarto render file.qmd -M title:"New title"`. The `-M` is a `pandoc` option.
6. ... can be set externally using `--metadata-file` but that does not seem to work now (c.f. #613). The `--metadata-file` is a `pandoc` option.
7. ... can be set externally with the help of [project scripts](https://quarto.org/docs/projects/quarto-projects.html#project-scripts) which `quarto` executes prior rendering documents. This requires using a [Quarto project](https://quarto.org/docs/projects/quarto-projects.html) feature.

Is the above a fair summary @cderv @jjallaire ?

**Reply by jjallaire** [reply]
This summary does seem correct but some of these limitations will soon be lifted by fixes we have in flight (cc @dragonstyle )

**Comment by RoelVerbelen** [comment]
Has there been any more recent development to enable using parameters in the YAML header?

Based on this discussion and [this related stack overflow question](https://stackoverflow.com/questions/77175146/quarto-using-parameters-in-report-title) I tried the following to no avail (using `quarto 1.4.549`):

``````
---
title: "`{python} my_parameter` {{< meta my_parameter >}} {{< meta params.my_parameter >}}"
jupyter:
  kernel: python3
---

```{python}
#| include: false
#| tags: [parameters]

my_parameter = 'test'
```

`{python} my_parameter`

{{< meta my_parameter >}} 

{{< meta params.my_parameter >}}
``````

![image](https://github.com/quarto-dev/quarto-cli/assets/12799126/b5f84e29-a708-4e10-95cb-f9031485a6d6)

**Reply by cderv** [reply]
YAML fields can't have inline code like that - it is parsed before execution happens, so it does not work. Also setting parameter for use with Jupyter is not something you can access in `meta` shortcode. Where did you see this usage ? This won't work either. 

Can you detail exactly your use case ? or Is the question generic ? 


Depending on what you want to do, there could be different solutions. But the `title` YAML field is quite specific as it must be known early for the processing of Quarto project. Having a real use case could help us show example

**Comment by RoelVerbelen** [comment]
Hi @cderv, thanks for your reply.

The use of `meta` shortcode in the YAML was claimed to work in [this answer](https://stackoverflow.com/a/77294225/3930480) on stackoverflow of the same question (in R context). 

In terms of my use cases: for data science projects, I set up exploratory data analysis reports which get run under various settings (e.g. changing the outcome variable, the data set (version) used, the filter settings) and like to see that reflected in the (sub)title of the report for immediate reference. Likewise, I build many different models and use the same reporting template for model evaluation. Elaborating on the latter with some code, using R and R markdown, I would run for example:

```r
rmarkdown::render(
  here("code", "freq-model-evaluation.Rmd"),
  output_file = glue(here("output", "html", "freq-model-evaluation-{model}-{version}.html")),
  params = list(
    actual = "claim_number",
    modelled = "modelled_freq",
    exposure = "exposure",
    model = "gbm",
    version = 1
  ),
  envir = new.env()
)
```

And the YAML header in my `.Rmd` would contain:

```YAML
---
title: "Frequency modelling" 
subtitle: "`r paste('Model evaluation for', params$model, 'version', params$version)`"
---
```

I'm now trying to set up a similar workflow in Python and Quarto.

Based on the `quarto.render()` code [here](https://github.com/quarto-dev/quarto-python/blob/main/quarto/render.py), I've set up a generic Python function to render a quarto report in an interactive terminal (as is convenient for interactive data science projects and for looping over various parameter settings easily within Python). Moving the title setting to this function (using the `-M title:"New title"` method) would technically work but not be an elegant solution as it decouples the title set in the YAML of the file itself and moves it to the generic Python quarto rendering function.

**Reply by cderv** [reply]
> The use of meta shortcode in the YAML

Meta shortcode in YAML should work. Inline R code should not. So you could fill subtitle using shortcode (`meta` or `env`, or `var` for project with variables yml file).  It should work with title to I believe. 

Did you try meta shortcode or other already ? 

````markdown
---
title: "{{< meta my_title >}}"
format: html
---

Content
````

then 
````powershell
quarto preview test.qmd --metadata my_title="filled"
````

This is working for me

**Reply by mcanouil** [reply]
Better than using inline in YAML, I suggest to produce YAML:

````qmd
---
title: "Frequency modelling" 
format: html
params:
  version: 1
  model: "the best"
engine: knitr
---

```{r}
#| output: asis
cat(
  "---",
  sprintf(
    "subtitle: Model evaluation for %s version %s\n",
    params[["model"]],
    params[["version"]]
  ),
  "---",
  sep = "\n"
)
```
````

You can also use `--metadata-file my-meta.yml`, see <https://github.com/mcanouil/demo-generate-quarto-invoices> as a demo.

The variable shortcode is also one way to make it work:
<img width="1624" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/d1426a32-3568-4d8b-8985-53dca2557add">

But be careful of where you are using shortcode as it might not work well for anchor links (https://github.com/quarto-dev/quarto-cli/issues/8546).
